### PR TITLE
Add `addFileExtension` function and `(<.>)` operator

### PR DESCRIPTION
**Pull request by ChShersh**
This PR add function that adds file extension according to #105. I'm not sure I did everything right because I didn't notice contribution policies. But I tried. Please, correct me, if I need to add/fix anything.

### CODE REVIEW

**Comment by NorfairKing** [comment]
I would also like some property tests: https://github.com/commercialhaskell/path/blob/master/test/ValidityTest.hs, but doesn't have to be in this PR.

File: src/Path.hs [hunk 1]
```diff
@@ -330,6 +332,41 @@ dirname (Path l) = Path (last (FilePath.splitPath l))
...
+-- >>> addFileExtension "symbols" $(mkRelFile "Data.List")
+-- "Data.List.symbols"
+-- >>> addFileExtension "evil/" $(mkRelFile "Data.List")
+-- *** Exception: InvalidRelFile "Data.List.evil/"
+--
+-- @since 0.6.1
+addFileExtension :: MonadThrow m
+  => String            -- ^ Extension to add
+  -> Path b File       -- ^ Old file name
+  -> m (Path b File)   -- ^ New file name with the desired extension added at the end
```
Comment (joehillen) [inline src/Path.hs]: Why is the argument order different than in [filepath](https://hackage.haskell.org/package/filepath-1.4.1.2/docs/System-FilePath-Posix.html#v:addExtension)? I think I prefer filepath's version.

File: src/Path.hs [hunk 2]
```diff
@@ -330,6 +332,41 @@ dirname (Path l) = Path (last (FilePath.splitPath l))
...
+-- | Add extension to given file path. Throws if the
+-- resulting filename does not parse.
+--
+-- >>> addFileExtension "symbols" $(mkRelFile "Data.List")
+-- "Data.List.symbols"
+-- >>> addFileExtension "evil/" $(mkRelFile "Data.List")
+-- *** Exception: InvalidRelFile "Data.List.evil/"
+--
+-- @since 0.6.1
+addFileExtension :: MonadThrow m
```
Comment (joehillen) [inline src/Path.hs]: Do we really need the word `File` in the function name? What is wrong with `addExtension`?

File: src/Path.hs (hunk 1 above)
Comment (NorfairKing) [inline src/Path.hs]: I disagree. This is easier to map.

Comment (joehillen) [inline src/Path.hs]: Good point.

File: src/Path.hs (hunk 2 above)
Comment (ChShersh) [inline src/Path.hs]: @joehillen I did it similar to `setFileExtension`. I thought consistency is better.

File: src/Path.hs [hunk 3]
```diff
@@ -330,6 +332,41 @@ dirname (Path l) = Path (last (FilePath.splitPath l))
...
+addFileExtension ext (Path path) =
+  if FilePath.isAbsolute path
+    then liftM coercePath (parseAbsFile (FilePath.addExtension path ext))
+    else liftM coercePath (parseRelFile (FilePath.addExtension path ext))
+  where coercePath :: Path a b -> Path a' b'
+        coercePath (Path a) = Path a
+
+-- | A synonym for 'addFileExtension' in the form of an operator.
+--
+-- >>> $(mkRelFile "Data.List") <.> "symbols"
```
Comment (ChShersh) [inline src/Path.hs]: @joehillen Sure, I can add more examples into comments.

File: src/Path.hs (hunk 2 above)
Comment (joehillen) [inline src/Path.hs]: Fair enough

File: src/Path.hs (hunk 3 above)
Comment (ChShersh) [inline src/Path.hs]: @joehillen @mrkkrp Btw, should I make examples expandable how it's done for [`Data.Bool.bool`](http://hackage.haskell.org/package/base-4.9.1.0/docs/Data-Bool.html#v:bool)? There will be more than couple examples so probably it makes sense to hide them by default.

Comment (mrkkrp) [inline src/Path.hs]: Personally, I like how it's done here: https://hackage.haskell.org/package/filepath-1.4.1.2/docs/System-FilePath-Posix.html#v:splitExtension, you can see all the subtle details of behavior right away, and I don't think it wastes too much space either. Use your judgement though.

Comment (joehillen) [inline src/Path.hs]: I also don't like hiding examples and prefer them front and center. Haddock's CSS make hidden examples easy to miss.

**Comment by ChShersh** [comment]
I've written examples not like in `filepath` library. I used `ghci`-like examples so later they can be easily verified and kept up-to-date using `doctest`. Probably adding `doctest` is a good idea and it's worth it.

**Comment by joehillen** [comment]
LGTM. @magthe or @NorfairKing either of you can merge it if it looks good to you.

I agree about adding doctest. It's been on my mental TODO list for a while.

**Pull request merged**